### PR TITLE
Introduced Parameter data class to encapsulate all information about single parameter of layer

### PR DESCRIPTION
**Pull request by knok16**

### CODE REVIEW

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import org.jetbrains.kotlinx.dl.api.core.KGraph
+import org.jetbrains.kotlinx.dl.api.core.initializer.Initializer
+import org.jetbrains.kotlinx.dl.api.core.regularizer.Regularizer
+import org.jetbrains.kotlinx.dl.api.core.util.getDType
+import org.tensorflow.Operand
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+import org.tensorflow.op.core.Variable
+
+public data class Parameter(
```
Comment (knok16) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:12]: Please advise on the naming

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    shape: Shape,
+    fanIn: Int,
+    fanOut: Int,
+    initializer: Initializer,
+    regularizer: Regularizer?
+): Parameter {
+    val tfVariable = tf.withName(variableName).variable(shape, getDType())
+
+    val initOp = initializer.apply(fanIn, fanOut, tf, tfVariable, variableName)
+    kGraph.addLayerVariable(tfVariable, isTrainable)
```
Comment (knok16) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:35]: Will go away later, when Trainable/Parametrized interfaces will be introduced for Layer

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt (hunk 1 above)
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:12]: At this moment it's not exactly Parameter.. I suggest to rename to KVariable

Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:12]: Need to add some KDocs here

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.jetbrains.kotlinx.dl.api.core.util.getDType
+import org.tensorflow.Operand
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+import org.tensorflow.op.core.Variable
+
+public data class Parameter(
+    val name: String,
+    val shape: Shape,
+    val tfVar: Variable<Float>,
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:15]: Please add KDocs and rename tfVar honestly to just variable

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt (hunk 2 above)
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:35]: Agree, it's not the goal of the current PR

File: examples/src/main/kotlin/examples/transferlearning/lenet/Example_6_Additional_training_and_new_dense_layers.kt [hunk 4]
```diff
@@ -37,12 +37,9 @@ fun additionalTrainingAndNewTopDenseLayers() {
...
     val layers = mutableListOf<Layer>()
     layers.add(input)
-    for (layer in otherLayers) {
-        if (layer is Conv2D || layer is MaxPool2D) {
-            layer.isTrainable = false
-            layers.add(layer)
-        }
-    }
+    layers.addAll(
+        otherLayers.filter { it is Conv2D || it is MaxPool2D }.onEach { if (it is Conv2D) it.isTrainable = false }
```
Comment (zaleslaw) [inline examples/src/main/kotlin/examples/transferlearning/lenet/Example_6_Additional_training_and_new_dense_layers.kt:41]: I'm not sure that after this refactoring the example became clear...

**Review by zaleslaw — CHANGES_REQUESTED** [review]

**Comment by juliabeliaeva** [comment]
Looks like changes in the examples are not related to the main goal of the PR and should be moved to a separate one.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/SeparableConv2D.kt [hunk 5]
```diff
@@ -119,45 +105,59 @@ public class SeparableConv2D(
         fanIn = (inputDepth * kernelSize[0] * kernelSize[1]).toInt()
         fanOut = ((outputDepth * kernelSize[0] * kernelSize[1] / (strides[0].toDouble() * strides[1])).roundToInt())
 
-        createSeparableConv2DVariables(tf, kGraph)
+        createSeparableConv2DVariables(tf, kGraph, numberOfChannels)
```
Comment (juliabeliaeva) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/SeparableConv2D.kt:108]: I would rather inline this method here. It seems to me that all the previous steps, like computing `numberOfChannels`, `inputDepth`/`outputDepth`, `fanIn`/`fanOut` are also the part of creating variables.

**Review by juliabeliaeva — CHANGES_REQUESTED** [review]

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    val initializer: Initializer?,
```
[the remark this answers is not in the document]
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:16]: Sorry, don't understand your comment, what do you suggest?

[the remark this answers is not in the document]
Comment (juliabeliaeva) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:16]: I suggest to remove `initializer` field.

[the remark this answers is not in the document]
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:16]: As I understand initializer/regularizer and so on will be removed from kGraph in the future PRs (looks like, the KVariable will be only one place where we could obtain initializer in the future)

[the remark this answers is not in the document]
Comment (juliabeliaeva) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:16]: KGraph does not keep `Initializer` instances, initializers are used once to get `initOp: Assign<Float>` and not needed after that (and `Parameter` has `initOp` already).

[the remark this answers is not in the document]
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/Parameter.kt:16]: Yes, I see now

**Comment by zaleslaw** [comment]
Hi, @knok16 do you have some time to read a review on this PR?

**Comment by zaleslaw** [comment]
Fixed in #324

**Pull request closed without merging**